### PR TITLE
Ask for feedback before deleting a WB instance to ask "why"

### PR DESCRIPTION
**Pull request by rosalieper**
Responses are stored and can be accessed as a file and the download should only be visible to internal users

### CODE REVIEW

File: tests/Routes/Wiki/DeletedWikiMetricsControllerTest.php [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        WikiManager::factory()->create( [
+            'wiki_id' => $wiki->id, 'user_id' => $user_id,
+        ] );
+        WikiSiteStats::factory()->create([
+            'wiki_id' => $wiki->id, 'pages' => 77, 'users' => $wiki_users
+        ]);
+        $wiki->created_at = $current_date->subWeeks($createdWeeksAgo);
+
+        $wiki->save();
+        WikiController::delete($this->get($this->route));
```
Comment (tarrow) [inline tests/Routes/Wiki/DeletedWikiMetricsControllerTest.php:60]: You probably meant to call this api using the test harness in a similar to what you did on line 40 in this patch. You can see a similar thing at tests/Routes/QsBatch/QsControllerTest.php#73 .

Comment (tarrow) [reply tests/Routes/Wiki/DeletedWikiMetricsControllerTest.php:60]: I think it could be better not to depend on the WikiController here. If the Controller method breaks then this test will break too and we will be confused.

Why don't we just run the same logic that would happen if we were to delete the wiki using the controller. Delete the Wiki and then have two cases: one where we add a reason and one where we don't?

File: tests/Jobs/DeleteWikiJobTest.php
Comment (tarrow) [inline tests/Jobs/DeleteWikiJobTest.php:1]: comment I appreciate this is a super long and confusing test which it would be nice to refactor. Maybe if you fancied doing this refactoring it would make sense to do it in a separate PR?

File: app/Http/Controllers/WikiController.php [hunk 2]
```diff
@@ -160,6 +161,7 @@ public function delete(Request $request): \Illuminate\Http\JsonResponse
         }
 
         // Delete the wiki
+        Wiki::find($wikiId)->update(['wiki_deletion_reason' => $wikiDeletionReason]);
```
[the remark this answers is not in the document]
Comment (tarrow) [reply app/Http/Controllers/WikiController.php:164]: Could we easily add a test for this?

File: app/Http/Middleware/AuthorisedUsersForDeletedWikiMetricsMiddleware.php [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace App\Http\Middleware;
+
+use Closure;
+use Illuminate\Http\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+class AuthorisedUsersForDeletedWikiMetricsMiddleware
+{
+    const WMDE_ALLOWED_USERS = [
```
[the remark this answers is not in the document]
Comment (tarrow) [reply app/Http/Middleware/AuthorisedUsersForDeletedWikiMetricsMiddleware.php:11]: yeah, I think that would make the most sense; maybe it could even be introduced first as a separate PR? "Add admin flag to user model" or something?

File: tests/Jobs/DeleteWikiJobTest.php
Comment (rosalieper) [inline tests/Jobs/DeleteWikiJobTest.php:1]: My IDE automatically remove trailing spaces from visited files. I tried disabling it in settings but it still does the change. I use phpstorm

File: app/Http/Controllers/WikiController.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (tarrow) [reply app/Http/Controllers/WikiController.php:179]: Did you decide it was too hard to have a test for the new behaviour we've added to this controller?

**Review by tarrow — COMMENTED** [review]
Generally looks good to me; nothing looks obviously glaringly out of place.

I could imagine that had we know this at the start we might have split this into say, 3, separate patches:
- Adding admin level functionality to the user class
- Add the changes to the WikiController to record the deletion reason
- Adding the new metrics controller to show the metrics to admin users.

I wanted to give you a chance to answer my question from before about adding some test coverage for the WikiController.

File: app/Http/Controllers/WikiController.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (rosalieper) [reply app/Http/Controllers/WikiController.php:179]: yeah, I could not think of a straight forward way for this

**Comment by rosalieper** [comment]
see the PR broken down in 3 here * https://github.com/wbstack/api/pull/803
* https://github.com/wbstack/api/pull/825
* https://github.com/wbstack/api/pull/826
